Савельев Григорий

luvairo-m opened the pull request:
@elybelyaev

luvairo-m committed:
Переработка исключений в стиле Result-паттерна.

luvairo-m committed:
Добавление тестов. Небольшие правки в TagsCloudVisualization.

elybelyaev commented on TagsCloud.Tests/TextAnalyzerTests.cs:
```diff
@@ -0,0 +1,36 @@
+            group.WordInfo.IsRussian.Should().Be(isEnglish);
+    }
+
+    private static IEnumerable<(WordTagGroup Group, bool isRussian)> GetTestGroups()
```
Похвально, что придумал сам как упростить написание тестов. Но существует встроенное в nunit решение, очень похожее на то, что ты тут предложил

https://docs.nunit.org/articles/nunit/writing-tests/TestCaseData.html

В чем преимущество - сейчас тест падает с ошибкой, и непонятно в какой из групп ошибка произошла. Если сделаешь с TestCaseSource+TestCaseData то ты увидишь, что ошибка упала в нескольких группах

elybelyaev replied:
Также ожидаю правки ошибочного поведения. Больше красных тестов у тебя нет

UPD: проблемы на моей стороне, кривая версия mystem

elybelyaev commented on TagsCloud.Tests/CloudOptionsBuilderTests.cs:
```diff
@@ -0,0 +1,52 @@
+
+[TestFixture]
+[TestOf(nameof(CloudOptionsBuilder))]
+public class CloudOptionsBuilderTests
```
В части тестов у тебя проверяется только поведение при ошибке. Но проверки на успешные сценарии тоже нужны, иначе как ты по юнит-тестам поймешь, что у тебя программа будет работать корректно?

elybelyaev commented on TagsCloud/Processors/InputProcessor.cs:
```diff
@@ -0,0 +1,126 @@
+
+    private Result<HashSet<WordTagGroup>> ExtractGroupsFromFile(string filename, IFileReader reader)
+    {
+        try
```
Кажется, тут try легко заменяется на паттерн Result: просто подумай какие ошибки могут возникнуть во всех типах reader и заранее отлови

Например, посредством стандартных библиотек довольно просто узнать о существовании файла до момента чтения из него

elybelyaev commented on TagsCloud/Processors/OutputProcessor.cs:
```diff
@@ -0,0 +1,36 @@
+
+    public Result<HashSet<WordTagGroup>> SaveVisualization(HashSet<WordTagGroup> wordGroups, string filename)
+    {
+        try
```
Аналогично с InputProcessor'ом

elybelyaev commented on TagsCloud/TextAnalysisTools/TextAnalyzer.cs:
```diff
@@ -0,0 +1,63 @@
+{
+    public static Result<HashSet<WordTagGroup>> FillWithAnalysis(HashSet<WordTagGroup> wordGroups)
+    {
+        try
```
А вот тут try оправдан, правда не круто что он оборачивает практически весь метод. Попробуй сделать так, чтобы все try в твоем коде были только там, где на более глубоких уровнях невозможно втащить паттерн Result

elybelyaev commented on TagsCloud/Extensions/ResultExtensions.cs:
```diff
@@ -0,0 +1,92 @@
+        return new Result<T>(error);
+    }
+
+    public static Result<T> Of<T>(Func<T> func, string error = null)
```
Кстати, этот метод (или какой-то из методов ниже) можно использовать при обращении с mystem

elybelyaev commented on TagsCloud/Contracts/ICloudProcessor.cs:
```diff
@@ -0,0 +1,11 @@
+using TagsCloud.Results;
+using TagsCloudVisualization;
+
+namespace TagsCloud.Contracts;
```
Относится ко всем интерфейсам

Обычно интерфейсы лежат где-то рядом с имплементацией, зачастую в одном неймспейсе. Иначе у тебя в разных частях программы будет постоянно юзинг одного и того же неймспейса TagsCloud.Contracts, да и сейчас довольно тяжело в них ориентироваться. У тебя здесь получилась небольшая свалка из всех интерфейсов что ты знаешь

Примерно то же можно сказать о экстеншенах, но для них это правило менее строго, часто можно увидеть все экстеншены в одном неймспейсе

Я не потребую исправлять это сейчас, просто подсвечиваю риски при дальнейшей разработке

luvairo-m committed:
Применение TestCaseSource + TestCaseData к тестам TextAnalyzer'a.
Перемещение интерфейсов из общей папки в папки реализаций.

luvairo-m committed:
Переработка TextAnalyzer: удаление try catch { } и добавление Result-логики.

luvairo-m committed:
Добавление валидатора путей в проект.
Расширение обработки ошибок работы с файлами и директориями.
Небольшие правки в тестах.

luvairo-m committed:
Добавление интеграционного теста.
Расширение тестов OptionsBuilder'ов положительными сценариями.
Небольшие правки в TagsCloud.
